pyomo/contrib/piecewise/triangulations.py

```diff
@@ -0,0 +1,717 @@
+        ]
+
+    for simplex_data in current_cube_path:
+        simplices.append(
+            get_one_j1_simplex(
+                current_v_0, simplex_data[1], simplex_data[0], 3, points_map
+            )
+        )
+
+    fix_vertices_incremental_order(simplices)
+    return simplices
+
+
+def _get_ordered_j1_triangulation_4d_and_above(points_map, num_pts, dim):
+    # step one: get a hamiltonian path in the appropriate grid graph (low-coordinate
+    # corners of the grid squares)
+    grid_hamiltonian = get_grid_hamiltonian(dim, num_pts)
+
+    # step 1.5: get a starting simplex. Anything that is *not* adjacent to the
+    # second square is fine. Since we always go from [0, ..., 0] to [0, ..., 1],
+    # i.e., j=`dim`, anything where `dim` is not the first or last symbol should
+    # always work. Let's stick it in the second place
+    start_perm = tuple([1] + [dim] + list(range(2, dim)))
+
+    # step two: for each square, get a sequence of simplices from a starting simplex,
+    # through the square, and then ending with a simplex adjacent to the next square.
+    # Then find the appropriate adjacent simplex to start on the next square
+    simplices = []
+    for i in range(len(grid_hamiltonian) - 1):
+        current_corner = grid_hamiltonian[i]
+        next_corner = grid_hamiltonian[i + 1]
+        # differing index
+        j = [k + 1 for k in range(dim) if current_corner[k] != next_corner[k]][0]
+        # border x_j value between this square and next
+        c = max(current_corner[j - 1], next_corner[j - 1])
+        v_0, sign = get_nearest_odd_and_sign_vec(current_corner)
+        # According to Todd, what we need is to end with a permutation where rho(n) = j
+        # if c is odd, and end with one where rho(1) = j if c is even. I think this
+        # is right -- basically the sign from the sign vector sometimes cancels
+        # out the sign from whether we are entering in the +c or -c direction.
+        if c % 2 == 0:
+            perm_sequence = get_Gn_hamiltonian(dim, start_perm, j, False)
+            for pi in perm_sequence:
+                simplices.append(get_one_j1_simplex(v_0, pi, sign, dim, points_map))
+        else:
+            perm_sequence = get_Gn_hamiltonian(dim, start_perm, j, True)
+            for pi in perm_sequence:
+                simplices.append(get_one_j1_simplex(v_0, pi, sign, dim, points_map))
+        # should be true regardless of odd or even
+        start_perm = perm_sequence[-1]
+
+    # step three: finish out the last square
+    # Any final permutation is fine; we are going nowhere after this
+    v_0, sign = get_nearest_odd_and_sign_vec(grid_hamiltonian[-1])
+    for pi in get_Gn_hamiltonian(dim, start_perm, 1, False):
+        simplices.append(get_one_j1_simplex(v_0, pi, sign, dim, points_map))
+
+    # fix vertices and return
+    fix_vertices_incremental_order(simplices)
+    return simplices
+
+
+def get_one_j1_simplex(v_0, pi, sign, dim, points_map):
```

Comment:
In general in this file, do you want to make all these helper functions public? You can, but then you might want to be a bit chivalrous and give them docstrings. Since we're in contrib I'm not so worried about backwards compatibility things at the moment, but it might be good to think through what might be general purpose and what's really just for here?

Reply:
I made most of these private